### DISCUSSION TITLE
Local Adjustments: User Interface

### DISCUSSION THREAD

**Discussion by rebio** [Beep6581/RawTherapee#6254] [Ideas]
## Intro
(Yeah, am sorry... Sometimes I'm kinda terrible at keeping it short.)

I saw a lot of discussions about Local Adjustments in the _Issues_ section regarding all kinds of different topics. One topic that I would like to go in a bit deeper with this discussion is the UI.

Personally, I think that the current UI works, but it is not where it could be - note that I'm not a photographer but more of an enthusiast. To me using the current UI feels more like using my computer with thick gloves. If you are a professional and think that the UI is perfect and is everything you need I hope you read my suggestions and drop a line.

My goal for this discussion is to get a UI (or ideas for a UI) that are both entry level friendly and easy to understand as well as adequate for the power user who wants/needs to turn on every nub there is. I don't think my designs are at their best either. When I thought through the designs they initially were OK in my head, but turned out to be more complicated than I wanted them to be. The goal would be to have a simple and self explanatory design that fits the rest of RT.

So all my suggestions are up for discussion and if you have the time or any idea for a self-drawn idea, I'm more than happy to see it. If you can not draw on your PC you can also use a pen and paper and scan or snapshot them with your smartphone. As long we can see what your idea is I don't really care about the image quality here.

Before we start: I want to be clear that I do not want to bash anybody or anybody's work here. I value your contribution and I really appreciate that all the contributers use their time and effort to improve this open source project!

## The current UI
In my opinion the UI looks like a patchwork that was created as the tools got ready and needed some sort of UI element to control them. There was no deeper masterplan or greater design philosophy in mind. That is not something bad, that's just how development of new functions work in general.

I want to focus on the UI with this discussion, but some elements just overlap. So there are some non-UI elements in here. A list of the current issues:
- Box in a box design
Everything looks very boxy. Why do we need a "Settings" box in the first place? If you add tools, there are even more boxes. And boxes within boxes. The first impression here for the 5.8 to 5.9 upgrader is not a clean looking UI, but more a convoluted picture.
- In the first line it says "Local Adjustments" and you have to turn it on to use it
The whole tab is turned off by default. I don't think that there is another tab that is handled that way. It should always be on by default (If there is no spot I wouldn't expect any delay when the image is processed). I also don't think that it needs to be named. There already is the tooltip that can be renamed.
- "Show additional settings" is off by default
It's not a good idea to disable it by default if you even need this option to change from an ellipse to a rectangle)
- Named buttons for options (Add, Delete, Rename, etc.)
If you can use a pictogram it should be preferred over text. It is smaller and there is no need for any translation. If the symbol is unclear there can be a small tooltip.
- Collapsable _Settings_ menu box
I don't quite get the use of this. Sure you can make more room for the tools - on the other hand you can scroll.
- Symmetric changes are hidden in some menu
- The working area preview is only active with one (and only one) tool active
Maybe this is not what "Preview ΔE" is about, but I would expect the button to show me the working area to ALWAYS show me the area that is effected. This preview should not be changed by the number of active tools (unless the tool changes the working area, which I think it shouldn't). I saw the tooltip of the "Preview ΔE"-button and I tried to work with the masking in a second tool. It did not work for me and I have no idea what I should do. I also think that it's not good if you have to fiddle around with the mask for every tool.
![RawTherapee - Local Adjustments -effect preview tooltip](https://user-images.githubusercontent.com/10911084/119240567-5602de00-bb48-11eb-8c7f-e879dbe4ab8a.png)
- Toolnames are not consistent throughout RawTherapee
Right now the tools in LA may have the same functionality as the ones in the rest of RT, but are combined (?) or have different names (?). My point here is that I can not say eg. _I want to add the L\*a\*b - Adjustments to this spot_ and just add the tools I already know from the rest of RT.
- Spot methode: Full image
Like... what? It's a Local Adjustment where the spot (that is used to determine the effected part of the selected area) determines what the actual area (eg. an ellipse) looke like? How does this make sense. I mean it would make sense as selectable "RT-spot shape", but not in in the _What type of spot are you_ setting.

## Suggestions and ideas
Here is an unsorted list of suggestions that again is kinda a mix of UI and usability
- Rename spot on double click or _Button + F2_
Right now F2 renames the file. Using Shift and Ctrl does the same. You can save a rather big button if you could rename the spot on double click or if you could use some shortcut. Pictograms would again help here.
- Show/Hide on double click
This would again save a big box that has no other use than do that rather limited task. The toggle could either be on double click or with a tickbox.
- Large Spot box
The box where the spots are listed is quite large and has a large unused fiel for columns. The box could grow with each new spot up to a certain point and be scrollable then like it is now when there are more than 6 spots.
- The preview of the effected area is always shown
Since it's very important to the entry level or casual user to reliably see the effected area this comes again. The effected area should always be shown regardless of whether one, more or no active tool is added to the list. Also there is no preview for the _Inverse_ mode.
- Move every other LA-like options to this section
It would be much better if all elements that effect only a part of the image would be collected here. Eg. the Graduated Filter or the mask of the Color Toning (this is also why I thought about stacked areas - you could eg. add a graduated filter and add a mask on the filter that excludes all the areas you don't want to effect)
- Area handles or spot can be moved freely
right now the spot is _fixed_ and the handles can only be moved along one axis. Often (eg. with heads) the ellipse shape is good, but the spot is not in the right position. The spot however can not be (freely) positioned somewhere where the handles of the outlined can pe positioned to fit the entire head without covering the background as well.
- Symmetrical shape size and rotation
Right now the symmetrical shape size is hidden in some submenu behind a dropdown bar - no newbee will ever find it there (especially since the UI controls are added right below the "Spot methode"). Being able to rotate the spot would also be helpfull. As hotkey Shift + drag could be used for symmetrical changes and Ctrl + drag could be used to rotate the spot.

## UI design examples
Even though I can see where my original concept is coming from, I think there is a large room for improvements. Here I present some ideas I had and tell you why I made the changes I did

### Local Adjustmets
![RawTherapee - Local Adjustments - Design example](https://user-images.githubusercontent.com/10911084/119270462-ae8fb500-bbf4-11eb-82f8-1193fc9bb329.png)

The changes:
- The ON/OFF button and the name was removed
As said, I don't think either of them are needed (unless you want to confuse new users when they try to figure out why their changes are not working)
- UI based on tabs
I quite like the existing tab design at the top so I tried to modify it for this usage. A tab here would be the collected changes in a certain area. There can be more than one tool active as well as more than one area active. Using the "-" and "+" you would remove and add tabs. You would start out with none or one tab. The tab's outline surrounds everything that changes either the effected area (ellipse, rectangle, etc.) or the effect itself (tools).
- Tab number 2
Here a second tab would exist that would have entirely different settings. To have a clear difference between the tab you are working in and every other one, the second tab's outline is much darker. Thinking about this, maybe the "2" could be a bit darker as well.
- Green line below the number of the tab
The green line is there because the original tab at the top has one and I did not want to make it blue to avoid confusion. Howener, this could may be used to quickly tell wether the tab is active (green) or disabled (red) with the UI element below
- UI elements right below the tab number
The symbols are used to (1) turn off/on the whole tab (2) duplicate the tab (3) invert the tab (will tools or the area be inverted? Well, I did not think about this until this moment...)
- Area box
The area box collects all elements that represent which area is effected. This could be a combination of multiple shape types at once to have some more complicated outlines that are beyond an ellipse and rectangles. But a box in a box? How dare I to criticise the original layout when I present this? Well, to be honest I just used the existing layout and did not put that much thought into this.
- Columns
"Type" is the area type. "Opacity" is the how much this element effects the entire area. "Spot" is just like the current spot, but listed in the box. "Name" is also the same. The dark background of the second element is the _active_ element that is changed by the UI elements below. The UI buttons at the bottom are to (1) add an area (2) remove the _active_ area (3) toggle the active area between 0% and it's last opacity value (4) duplicate the active area (5) rename the active area (6) invert the active area.
- Spot size
This is what it is doing right now. Actually more to show that I'm not against hiding all config tools in the next section.
- Area settings
Here you would find everything that is accessed when you click on "Show additional settings" now (minus the tools that can be accessed in the boxy list above). The icon on the right side is to show the working area.
- Tools
The tools are listed like they are now. They are listed as they are now since this feels very much like the rest of RT. The "+" and "-" buttons would add and remove tools.

Here is the SVG file of this design example. Note that github does not allow SVG files to be uploaded. So I renamed it to a jpg file. If you want to view it as SVG again you have to download it and change the file extension!
![SVG of RawTherapee - Local Adjustments - add tool design example](https://user-images.githubusercontent.com/10911084/119274004-a4c27d80-bc05-11eb-9ba6-802de125895b.jpg)

#### The part where I question my own design
A list of thoughts/doubts I have (a question to professional end users)
- How many tabs/areas would you really need. How many do you usually use in one photograph?
The main idea why I came up with combined smaller areas to add one larger area is because this can be done rather easy in Lightroom. IMO the level of adjustability is rather small right now. I think 3 was the largest ammount of local adjustment fields I used ever (but I also only started to use RT about a year ago and only about 9 months ago I discovered the nightly builds so what do I know)
- Toggle tabs, areas and tools
Would it be usefull to be able to toggle all of them? Or any of them? I was hoping for a way to easily disable a spot for a moment without having to delete the actual spot. Now there is and yet it is only a tiny box below the spot-box list and it is not shown in the spot-box list even there would be a column free for that,
- Wouldn't it be increadibly bothersome to create a new tab just to add a radient filter?
If you want to do different changes to two different areas you would need a new tab for each change. Only more effected area can be collected in one tab.
- A single spot for the entire area
If we could have multiple combined areas anyway, why would you have a dedicated spot for each area?
Why not define an area and define one spot. Both could be set without being linked to each other beside being effecting the same tab
- Should really be all other area settings be hidden in the menu?
Aren't there some useful sliders that you use every time?

### Adding a tool
![RawTherapee - Local Adjustments - add tool design example](https://user-images.githubusercontent.com/10911084/119272973-a9d0fe00-bc00-11eb-9a0d-c5599b358a0d.png)
A second concept to think about: How a tool is added.
- When you add a tool it opens up kind a mirror of the top tab bar
You would be able to find every tool where you would find it in the top bar. There are no renamed tools. There are no tool combined into one. There is nothing that you would have to think about if you know the rest of RT. In this example you are about to select a tool from the Exposure Tab.

#### The part where I again question my own design
A list of thoughts/doubts I have (a question to professional end users)
- Bulky and boxy design
Instead of remembering a few toolnames and what they do, you would have to traverse through a boxy and packed menu with submenus. Who in their right mind would suggest such a design for a productive use? Would this be any helpful?
- Removing tools
At some point I read that there are some duplicate tools or tools that do roughly the same. Why would we need to add a second layer to this problem when the tools are added yet another duplicate time.

## Other ideas
Some random thoughts and ideas:
- Some of the buttons have names that are not easy to understand if you are not a programmer or a photographer. It should be somewhat clear what it does, but what does "Preview ΔE" do?
- Some area tool that let's you paint on out an area would be nice. This could make selecting an area with a more complicated outline much faster
- Why is the mask green?
The mask of the Color Toning is yellow and usually masks are black and white. I'm not oppose a colour to seperate it from dark and bright backgrounds, but why yet another colour?

## This can be discussed when 5.9 is released
Can it though?

Wouldn't it be nice to have the version after 5.9 open the same pp3 file that 5.9 saved and work exactly like 5.9 did? I'm not oppose publishing a new version, but I think the UI could be a bit more polished before it's release. Otherwise users would maybe have to relearn how this tool is used and where to find stuff (it's something that I really dislike with every software if they change the UI for every version).

In any case I would like to hear what you think about my ideas and I would also like to hear yours!

**Comment by Thanatomanic** [comment]
@rebio Thanks for taking the time to write this! I'll write a more in-depth reply soon(ish).

> I'm not oppose publishing a new version, but I think the UI could be a bit more polished before it's release. Otherwise users would maybe have to relearn how this tool is used and where to find stuff (it's something that I really dislike with every software if they change the UI for every version).

This is a hard 'no, not before 5.9' for me. There is a trade-off between many things. Ys, we can certainly polish the UI, but it is now more important to get a solid release out. Since all functionality is working, we should be good to go. Optimization or changing the interface will delay this (much) more, because we're not the most skilled in that.
As to your second argument, I partly agree. It would be optimal to publish a tool without ever having to change its interface. However, if the changes improve the UX well enough, the effort to relearn a tool should be minimal and the net effect should be very positive. I think we should aim at that for possible future change.

**Comment by danielsz** [comment]
I like the UI, but I cannot figure out how to reorder the adjustment layers. Often I want to push at the bottom a full-screen layer with exposure adjustments. After I already do spot layers.

**Reply by Lawrence37** [reply]
It's not possible to rearrange layers at the moment.
